engine/triangulation/dim3/surfaces.cpp

```diff
@@ -237,35 +237,85 @@ bool Triangulation<3>::isZeroEfficient() {
+    std::set<Edge<3>*> disjoint;
+    std::set<Edge<3>*> intersecting;
 
-    // Run through all vertex surfaces.
-    unsigned long nSurfaces = surfaces->size();
-    const NormalSurface* s;
-    LargeInteger chi;
-    for (unsigned long i = 0; i < nSurfaces; i++) {
-        s = surfaces->surface(i);
+    // We also keep track of each edge e that is not yet assumed disjoint but that is a candidate for this assumption.
+    std::list<Edge<3>*> candidate_disjoint;
+    Tetrahedron<3>* tet = simplex(0);
+        
+    // At the outset, we may regard any edge as a candidate.
+    // We will do so for each of the three edges in tet incident to vertex 3.
+    // The triangulation is connected.
+    // So these exhaust the possibilities for a splitting surface.
+
+    for (int i = 0; i < 3; i++){
+        candidate_disjoint.clear();
+        disjoint.clear();
+        intersecting.clear();
+
+        // Outset
+        Edge<3>* ei3 = tet->edge(i,3);
+        candidate_disjoint.push_front(ei3);
+        bool obstructed = false;
 
-        if (s->isSplitting()){
-            splittingSurface_ = true;
-            break;
+        // Main loop
+        while (!candidate_disjoint.empty()){
+            Edge<3>* e = candidate_disjoint.back();
+            for (auto& emb : *e){
+                Tetrahedron<3>* tet_e = emb.simplex();
+                Perm<4> v_perm = emb.vertices();
+
+                // The splitting surface should intersect the lateral edges of the candidate edge in the given embedding.
+                Edge<3>* lat02 = tet_e->edge(v_perm[0],v_perm[2]);
+                Edge<3>* lat03 = tet_e->edge(v_perm[0],v_perm[3]);
+                Edge<3>* lat12 = tet_e->edge(v_perm[1],v_perm[2]);
+                Edge<3>* lat13 = tet_e->edge(v_perm[1],v_perm[3]);
+                intersecting.insert(lat02);
+                intersecting.insert(lat03);
+                intersecting.insert(lat12);
+                intersecting.insert(lat13);
+            }
+
+            // Now we check for a local obstruction to a splitting surface opposite e.
+            if (intersecting.count(e))
+                break; // without removing e
+            else {
+                candidate_disjoint.pop_back(); // remove e
+                disjoint.insert(e);
+                // Regard the edges opposite e as candidates if they're not already assumed disjoint.
+                for (auto& emb : *e){
+                    Tetrahedron<3>* tet_e = emb.simplex();
+                    Perm<4> v_perm = emb.vertices();
+                    Edge<3>* opp = tet_e->edge(v_perm[2],v_perm[3]);
+                    if (!disjoint.count(opp)) candidate_disjoint.push_front(opp);
+                }
+            }
         }
+        if (candidate_disjoint.empty())
+            // We didn't break without removing an edge.
+            // So we must have partitioned the edges into disjoint and intersecting.
+            // Thus there is a splitting surface.
+            return true;
```

Comment:
Is there a reason for not setting splittingSurface_ = true here?

Reply:
I did set ``splittingSurface_ = true`` there; looks like the pull request is outdated for some reason. However, I made that mistake earlier in the routine; I'll fix it.

Reply:
Feel free to close this PR; I'll submit the next one to the 6.0_pre branch.